packages/components/src/components/Calculator/CalculatorInput.tsx

```diff
@@ -0,0 +1,57 @@
+import { FC } from "react";
+import { useFormContext } from "react-hook-form";
+
+import { SqInput } from "@quri/squiggle-lang";
+import {
+  CheckboxFormField,
+  SelectStringFormField,
+  TextAreaFormField,
+  TextFormField,
+} from "@quri/ui";
+
+import { PlaygroundSettings } from "../PlaygroundSettings.js";
+import { ValueResultViewer } from "./ValueResultViewer.js";
+import { SqValueResult } from "./index.js";
+
+export const CalculatorInput: FC<{
+  input: SqInput;
+  result: SqValueResult | undefined;
+  settings: PlaygroundSettings;
+}> = ({ input, result, settings }) => {
+  const form = useFormContext();
+  const { name, description } = input;
+
+  const code = form.getValues(input.name);
+
+  return (
+    <div className="flex flex-col mb-2">
+      <div className="text-sm font-medium text-gray-800">{name}</div>
+      {description && (
+        <div className="text-sm text-gray-400">{description}</div>
+      )}
+
+      <div className="flex-grow mt-1 max-w-xs">
+        {input.tag === "text" && (
+          <TextFormField
```

Comment:
I find it really magic-heavy that we don't need any hooks to these. Pretty uncomfortable about that, though I realize this is how the FormFields work. Curious if there's some more explicit option, though maybe not for this PR.

Reply:
I agree in general that explicit is better than implicit, but I think this is the right approach for forms, because otherwise we'd have to write 2x-3x as much code.

For more control, you can use the generic FormField or ControlledFormField (or FormInput and ControlledFormInput if you don't need label & description), it's documented here: https://quri-ui.vercel.app/?path=/docs/ui-forms--docs#generic-field-components.

Reply:
Yea, that could be quite fair here. It's a pity things are this way, but forms are just pretty bad.